keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/ec2/ClusterDependencyContainer.kt

```diff
@@ -0,0 +1,68 @@
+package com.netflix.spinnaker.keel.api.ec2
+
+import com.netflix.spinnaker.keel.api.Locatable
+import com.netflix.spinnaker.keel.api.ResourceSpec
+import com.netflix.spinnaker.keel.api.SubnetAwareLocations
+import com.netflix.spinnaker.keel.core.api.ClusterDependencies
+
+/**
+ * A [ResourceSpec] that has default and override cluster dependencies (i.e. security groups, load
+ * balancers, and/or target groups).
+ */
+interface OverrideableClusterDependencyContainer<T : ClusterDependencyContainer> :
+  Locatable<SubnetAwareLocations> {
+  val defaults: T
+
+  /**
+   * Overrides by region. Map keys are region names.
+   */
+  val overrides: Map<String, T>
+}
+
+interface ClusterDependencyContainer {
+  val dependencies: ClusterDependencies?
+}
+
+/**
+ * Resolves overrides and returns a map of security group name to the regions it is required in.
+ * For example:
+ * ```
+ * fnord -> (us-east-1, us-west-2)
+ * fnord-elb -> (us-east-1, us-west-2)
+ * fnord-ext -> (us-east-1)
+ * ```
+ */
+val OverrideableClusterDependencyContainer<*>.securityGroupsByRegion: Map<String, Set<String>>
+  get() = dependencyByRegion { it.securityGroupNames }
+
+val OverrideableClusterDependencyContainer<*>.loadBalancersByRegion: Map<String, Set<String>>
+  get() = dependencyByRegion { it.loadBalancerNames }
+
+val OverrideableClusterDependencyContainer<*>.targetGroupsByRegion: Map<String, Set<String>>
+  get() = dependencyByRegion { it.targetGroups }
+
+private fun OverrideableClusterDependencyContainer<*>.dependencyByRegion(fn: (ClusterDependencies) -> Set<String>): Map<String, Set<String>> {
+  val regions = locations.regions.map { it.name }.toSet()
+  val defaults = (defaults.dependencies?.let(fn) ?: emptySet())
+    .associateWith { regions }
+  val overrides = overrides
+    .map { (region, spec) ->
+      spec.dependencies?.let(fn)?.associateWith { setOf(region) } ?: emptyMap()
+    }
+    .merge()
+  return listOf(defaults, overrides)
+    .merge()
+}
+
+/**
+ * Reduces a collection of `Map<*, Set>` down to a single map where the values contain the combined
+ * set of values from all elements in the original collection with a common key.
+ */
```

Comment:
I feel like the things in this file could benefit from some typing. It's pretty hard for me to reason about `Map<String, Set<String>>` etc. Could you create a data class to make working with this a little easier?

Reply:
I think that would really only be able to replace the `Map` part but I'll give it a shot & see if it makes things clearer.